Add readonly typings for Set and Map

gmoothart opened the pull request:
Fixes #10919

Similar to ReadonlyArray, these typings remove the mutation methods from Set and Map so that they can only be initialized by the constructor.

[Example](https://www.typescriptlang.org/play/#src=interface%20ReadonlyMap%3CK%2C%20V%3E%20%7B%0D%0A%20%20forEach(callbackfn%3A%20(value%3A%20V%2C%20index%3A%20K%2C%20map%3A%20ReadonlyMap%3CK%2C%20V%3E)%20%3D%3E%20void%2C%20thisArg%3F%3A%20any)%3A%20void%3B%0D%0A%20%20get(key%3A%20K)%3A%20V%7Cundefined%3B%0D%0A%20%20has(key%3A%20K)%3A%20boolean%3B%0D%0A%20%20readonly%20size%3A%20number%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20ReadonlySet%3CT%3E%20%7B%0D%0A%20%20forEach(callbackfn%3A%20(value%3A%20T%2C%20index%3A%20T%2C%20set%3A%20ReadonlySet%3CT%3E)%20%3D%3E%20void%2C%20thisArg%3F%3A%20any)%3A%0D%0A%20%20%20%20%20%20void%3B%0D%0A%20%20has(value%3A%20T)%3A%20boolean%3B%0D%0A%20%20readonly%20size%3A%20number%3B%0D%0A%7D%0D%0A%0D%0Alet%20roMap%3A%20ReadonlyMap%3Cstring%2C%20string%3E%20%3D%20new%20Map(%0D%0A%09%5B%5B%22key1%22%2C%20%22value1%22%5D%2C%20%5B%22key2%22%2C%20%22value2%22%5D%5D)%3B%0D%0Alet%20roSet%3A%20ReadonlySet%3Cnumber%3E%20%3D%20new%20Set(%5B1%2C2%2C3%2C4%2C4%5D)%3B%0D%0A%0D%0A%0D%0Alet%20v1%20%3D%20roMap.get(%22key1%22)%3B%0D%0A%2F%2F%20not%20available%3A%0D%0A%2F%2FroMap.set(%22key1%22%2C%20%22foo%22)%3B%0D%0A%0D%0Alet%20b%20%3D%20roSet.has(5)%3B%0D%0A%2F%2F%20not%20available%3A%0D%0A%2F%2FroSet.delete(4)%3B%0D%0A)

msftclas commented:
Hi __@gmoothart__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.
    </span>

TTYL, MSBOT;

mhegazy commented:
thanks @gmoothart. can you please sign the CLA?

mhegazy commented on src/lib/es2015.collection.d.ts:
```diff
@@ -15,6 +15,15 @@ interface MapConstructor {
+      thisArg?: any): void;
+  get(key: K): V|undefined;
+  has(key: K): boolean;
+  readonly size: number;
```
Please match the formatting for the rest of the file, that is 4 spaces.

mhegazy commented on src/lib/es2015.collection.d.ts:
```diff
@@ -15,6 +15,15 @@ interface MapConstructor {
+  forEach(
+      callbackfn: (value: V, key: K, map: ReadonlyMap<K, V>) => void,
+      thisArg?: any): void;
+  get(key: K): V|undefined;
```
and the spacing as well.

mhegazy commented:
I guess for the CLA i need to talk to @naomiblack.

gmoothart commented:
Formatting fixed. I also updated some parameters names which were wrong, and fixed the example.

It would be nice to be able to have type inference for the generic parameters, but I don't think it's possible in this case?

Re: the CLA, there is a custom Google/Microsoft one that I did sign and yeah, Naomi should be able to help you.
